### ISSUE TITLE
Attribute statements with parentheses not correctly parsed

### ISSUE THREAD

**Issue by pheibarrass** [cmacmackin/ford#76]
Attribute statements for module entities are not correctly parsed when they contain parentheses.  For example:

    module a
      allocatable i(:)
      dimension j(1)
      parameter (k=1)
      pointer l(:)
    end module a

is rejected with

> Exception: Found procedure call in MODULE

**Comment by cmacmackin** [comment]
The reason that this is rejected is because you are implicitly declaring your variables. FORD does not support this, with the exception of implicitly declared dummy-arguments in procedures. As such, FORD is misinterpreting the lines in this example as procedure calls, which are not allowed in the declaration area of a module.

I was not aware that the allocatable, dimension, and pointer attributes could be applied via these sorts of statements, although I now see that they are there in the standard. In a future release I will add code to support them and, in turn, prevent these sorts of errors. I don't plan to support implicit typing, however.

**Comment by pheibarrass** [comment]
It probably changes nothing, but If I change the variables to be explicitly typed (with `implicit none; integer i,j,k,l` at the start) the problem still comes up.

**Comment by cmacmackin** [comment]
Well, there are really two components to this issue: implicit typing and the attribute statements. It's really the attribute statements which are causing the crash, although lack of support for explicit typing means that even if attribute statements were supported you wouldn't get the results you expect.

For the moment, to make FORD run you'll need something like
```fortran
module a
  integer, allocatable:: i(:)
  integer, dimension(1) :: j     ! or integer j(1)
  integer, parameter :: k=1
  integer, pointer :: l(:)
end module a
```

**Comment by pheibarrass** [comment]
That workaround (which is good style, anyway) is good.

I noticed, although the resulting error is different, that there is also a problem with the `external` attribute statement.  Again, for functions one could do `integer, external :: ifun`, but for subroutines there isn't that option.

Happy to submit that as a separate issue, if you'd prefer.

**Comment by cmacmackin** [comment]
Could you give me an example of code that is failing? This sounds like 
another case of me not being aware of all of the ways in which these 
statements can be used. Because there are so many things which can be 
mistaken for procedure calls (many of which I don't know exist) and I 
don't know when I'll be able to support all of them I think I'll think 
this stop causing a run-time crash.

On 26/08/15 04:39 PM, pheibarrass wrote:
>
> That workaround (which is good style, anyway) is good.
>
> I noticed, although the resulting error is different, that there is 
> also a problem with the |external| attribute statement. Again, for 
> functions one could do |integer, external :: ifun|, but for 
> subroutines there isn't that option.
>
> Happy to submit that as a separate issue, if you'd prefer.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cmacmackin/ford/issues/76#issuecomment-135148884>.
>

-- 
Chris MacMackin
Saint Mary's University
Curriculum Vitae <http://ap.smu.ca/%7Ecmacmack/CV.pdf>

**Comment by pheibarrass** [comment]
module a
      external extsub
    end module a

complains that

> Exception: EXTERNAL declaration in MODULE

As before, the external attribute may be applied to a function on its declaration

    integer, external :: extfun

but not with an external statement.

**Comment by cmacmackin** [comment]
I've fixed this issue, along with the ones for function suffixes and deferred type-bound procedures. I've pushed the latest version of the code to GitHub. It would be greatly appreciated if you could try this out on your codes which were causing errors before I release it. I'm concerned that I might not have anticipated all of the potential pitfalls in my various toy test-cases, so testing on real code is always useful.

**Comment by pheibarrass** [comment]
Thanks, those various bits seem to be working now.

One minor thing, though: it looks like there's an extra "(" in the argument intent list.

**Comment by cmacmackin** [comment]
Fixed that. All fixes available in [version 4.2.2](https://github.com/cmacmackin/ford/releases/tag/v4.2.2).
